File: Graphs/UndirectedGraph/source/undirected_graph.js
class UndirectedGraph {
    constructor() {
        this.adjacencyList = {};
    }

    addVertex(vertex) {
        if (!this.adjacencyList[vertex]) {
            this.adjacencyList[vertex] = [];
        }
    }

    addEdge(vertex1, vertex2) {
        if (!this.adjacencyList[vertex1]) {
            this.addVertex(vertex1);
        }
        if (!this.adjacencyList[vertex2]) {
            this.addVertex(vertex2);
        }

        this.adjacencyList[vertex1].push(vertex2);
        this.adjacencyList[vertex2].push(vertex1);
    }

    hasEdge(source, target) {
        return this.adjacencyList[source]?.includes(target) || false;
    }

    print() {
        for (const [vertex, neighbors] of Object.entries(this.adjacencyList)) {
            console.log(vertex, " : ", neighbors);
        }
    }
}

File: Graphs/UndirectedGraph/source/undirected_graph.py
class UndirectedGraph:
    def __init__(self):
        self.adjacency_list = {}

    def add_vertex(self, vertex):
        if vertex not in self.adjacency_list:
            self.adjacency_list[vertex] = []

    def add_edge(self, vertex1, vertex2):
        if vertex1 not in self.adjacency_list:
            self.add_vertex(vertex1)
        if vertex2 not in self.adjacency_list:
            self.add_edge(vertex2)

        # Add the edge in both directions (bidirectional)
        self.adjacency_list[vertex1].append(vertex2)
        self.adjacency_list[vertex2].append(vertex1)

    def has_edge(self, source, target):
        return target in self.adjacency_list.get(source, [])

    def print(self):
        for vertex, neighbors in self.adjacency_list.items():
            print(f"{vertex} : {neighbors}")
